File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int n;
    static int[][] board;

    static int blue = 0;
    static int white = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());

        board = new int[n][n];

        StringTokenizer st;
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        divide(0, 0, n);

        System.out.println(white);
        System.out.println(blue);
    }

    private static void divide(int row, int col, int size) {

        if (isSameColor(row, col, size)) {
            if (board[row][col] == 0) {
                white++;
            } else {
                blue++;
            }
            return;
        }

        int half = size / 2;

        divide(row, col, half);
        divide(row + half, col, half);
        divide(row, col + half, half);
        divide(row + half, col + half, half);
    }

    private static boolean isSameColor(int row, int col, int size) {
        int curColor = board[row][col];

        for (int i = row; i < row + size; i++) {
            for (int j = col; j < col + size; j++) {
                if (board[i][j] != curColor) {
                    return false;
                }
            }
        }

        return true;
    }


}

File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.kt
import java.util.StringTokenizer

fun main() {
    val n = readln().toInt()
    val paper = Array(n) { IntArray(n) }
    var white = 0
    var blue = 0

    for (i in paper.indices) {
        val st = StringTokenizer(readln())
        for (j in paper[i].indices) {
            paper[i][j] = st.nextToken().toInt()
        }
    }

    fun cut(row: Int, col: Int, size: Int) {

        fun isSameColor(row: Int, col: Int, size: Int): Boolean {
            val curColor = paper[row][col]

            for (i in row until row + size) {
                for (j in col until col + size) {
                    if (paper[i][j] != curColor) {
                        return false
                    }
                }
            }

            return true
        }

        if (isSameColor(row, col, size)) {
            if (paper[row][col] == 0) {
                white++
            } else {
                blue++
            }
            return
        }

        val half = size / 2

        cut(row, col, half)
        cut(row + half, col, half)
        cut(row, col + half, half)
        cut(row + half, col + half, half)
    }

    cut(0, 0, n)

    println(white)
    println(blue)
}